torch/_decomp/decompositions.py

```diff
@@ -1983,6 +1983,135 @@ def get_scale_value(scales, idx):
+    indices = []
+    num_spatial_dims = len(output_size)
+    for d in range(num_spatial_dims):
+        # Math matches aten/src/ATen/native/cpu/UpSampleKernel.cpp
+        # Indices are computed as following:
+        # scale = isize / osize
+        # input_index = floor(output_index * scale)
+        # Same as OpenCV INTER_NEAREST
+        osize = sym_float(output_size[d])
+        output_indices = torch.arange(
+            sym_int(osize), dtype=input.dtype, device=input.device
+        )
+        isize = sym_float(input.shape[-num_spatial_dims + d])
+        scale = isize / osize
+        input_indices = torch.floor(output_indices * scale).to(torch.int64)
+        for _ in range(num_spatial_dims - 1 - d):
+            input_indices = input_indices.unsqueeze(-1)
+        indices.append(input_indices)
+    return tuple(indices)
+
+
+@register_decomposition(torch.ops.aten.upsample_nearest1d.default)
+@torch.ops.aten.upsample_nearest1d.default.py_impl(DispatchKey.Autograd)
+@pw_cast_for_opmath
+def upsample_nearest1d(
+    input: Tensor,
+    output_size: List[Union[int, float]],
+    scales: Optional[float] = None,
+) -> Tensor:
+    (l_indices,) = _compute_upsample_nearest_indices(input, output_size)
+    result = input[:, :, l_indices]
+    return result
+
+
+@register_decomposition(torch.ops.aten.upsample_nearest2d.default)
+@torch.ops.aten.upsample_nearest2d.default.py_impl(DispatchKey.Autograd)
+@pw_cast_for_opmath
+def upsample_nearest2d(
+    input: Tensor,
+    output_size: List[Union[int, float]],
+    scales_h: Optional[float] = None,
+    scales_w: Optional[float] = None,
+) -> Tensor:
+    h_indices, w_indices = _compute_upsample_nearest_indices(input, output_size)
+    result = input[:, :, h_indices, w_indices]
+
+    # convert output to correct memory format, if necessary
+    memory_format = utils.suggest_memory_format(input)
+
+    # following "heuristic: only use channels_last path when it's faster than the contiguous path"
+    _, n_channels, _, _ = input.shape
+    if input.device.type == "cuda" and n_channels < 16:
+        memory_format = torch.contiguous_format
+
+    result = result.contiguous(memory_format=memory_format)
+
+    return result
+
+
+@register_decomposition(torch.ops.aten.upsample_nearest3d.default)
+@torch.ops.aten.upsample_nearest3d.default.py_impl(DispatchKey.Autograd)
+@pw_cast_for_opmath
+def upsample_nearest3d(
+    input: Tensor,
+    output_size: List[Union[int, float]],
+    scales_d: Optional[float] = None,
+    scales_h: Optional[float] = None,
+    scales_w: Optional[float] = None,
+) -> Tensor:
+    d_indices, h_indices, w_indices = _compute_upsample_nearest_indices(
+        input, output_size
+    )
+    result = input[:, :, d_indices, h_indices, w_indices]
+
+    # convert output to correct memory format, if necessary
+    memory_format = utils.suggest_memory_format(input)
+
+    # following "heuristic: only use channels_last path when it's faster than the contiguous path"
```

Comment:
this is only for the 2d implementation. [here](https://github.com/pytorch/pytorch/blob/07c61685c81820fa9c9d74cf879700d896bc49c3/aten/src/ATen/native/cuda/UpSampleNearest2d.cu#L239) and [here](https://github.com/pytorch/pytorch/blob/07c61685c81820fa9c9d74cf879700d896bc49c3/aten/src/ATen/native/cuda/UpSampleBilinear2d.cu#L288)

Reply:
Cool, wasn't sure if there was an analogous channels_last3d thing going on. Will remove